torch/csrc/jit/codegen/cuda/kernel_cache.h

```diff
@@ -335,6 +361,82 @@ class TORCH_CUDA_CU_API InputsIdLookup : public NonCopyable {
   std::unordered_map<std::string, EncodingEntry> encoding_lookup_;
 };
 
+//! OptimizedFusionGraph is another layer of caching representing
+//!  input dependent segmentation and any optimizations done before
+//!  segmentation.
+//! The original input id based shortcut to the compiled
+//!  kernels is still maintained for exactly repeated inputs.
+//!  For new input sizes, a canUseFor interface is provided to check if this
+//!  cached optimized graph is compatible with the given input.
+class TORCH_CUDA_CU_API OptimizedFusionGraph {
+ public:
+  explicit OptimizedFusionGraph(
+      Fusion* fusion,
+      const at::ArrayRef<IValue>& inputs);
+
+  std::vector<at::Tensor> runFusionWithInputs(
+      const at::ArrayRef<IValue>& inputs,
+      size_t cache_id) {
+    auto fusion_kernel_runtime =
+        fusion_kernel_runtime_cache_.getRt(inputs, cache_id);
+    // Propagate the unique_id so the contained fusionExecutors in the runtime
+    //  entry will cache the buffer sizes and launch params based on this id.
+
+    auto&& ret = fusion_kernel_runtime->runWithInput(inputs, cache_id);
+    if (profiling_) {
+      most_recent_executor_log_ =
+          fusion_kernel_runtime_cache_.getMostRecentExecutorLog();
+    }
+    return std::move(ret);
+  }
+
+  ExecutorLog getMostRecentExecutorInfo() {
+    TORCH_INTERNAL_ASSERT(!isSegmented());
+    return most_recent_executor_log_;
+  }
+
+  void evictId(size_t cache_id) {
+    fusion_kernel_runtime_cache_.evictId(cache_id);
+  }
+
+  //! Checks if this graph can be re-used for the new given input,
+  //!  points to check here should include,
+  //!   1. Pre-segmentation optimizations are still valid, such as
+  //!    welford translation
+  //!   2. The segmented groups are still valid and can be scheduled
+  bool canReUseFor(const at::ArrayRef<IValue>& inputs);
```

Comment:
I don't know that I understand if this is the correct check. canSchedule is not sufficient to know if we should reuse a kernel. Only if the heuristics struct matches do we know if we can reuse it.

Reply:
Sorry if this is confusing due to an extra layer of caching. This function actually doesn't go straight into deciding we already have a kernel to run for the inputs. It only decides if we need to re-run segmentation. The heuristics parameters are maintained in the next level and after passing `canReUseFor` we can guarantee that we can get a vector of heuristic structs for the new input based on the current segmented fusion and there we can check if we already have a re-usable kernel or we need to compile a new one.